pythran/pythonic/numpy/correlate.hpp

```diff
@@ -0,0 +1,182 @@
+#include "pythonic/numpy/asarray.hpp"
+#include "pythonic/types/ndarray.hpp"
+
+#if defined(PYTHRAN_BLAS_ATLAS) || defined(PYTHRAN_BLAS_SATLAS)
+extern "C" {
+#endif
+#include <cblas.h>
+#if defined(PYTHRAN_BLAS_ATLAS) || defined(PYTHRAN_BLAS_SATLAS)
+}
+#endif
+
+PYTHONIC_NS_BEGIN
+
+#define MIN(A, B) (((A) < (B)) ? (A) : (B))
+#define MAX(A, B) (((A) > (B)) ? (A) : (B))
+
+namespace numpy
+{
+
+  // no_conj is unused here, but needed for the constant signature
+  float dot(int N, float *a, float *b, float no_conj)
+  {
+    return cblas_sdot(N, a, 1, b, 1);
+  }
+  double dot(int N, double *a, double *b, float no_conj)
+  {
+    return cblas_ddot(N, a, 1, b, 1);
+  }
+
+  // versions for which we don't have Blas support
+  double dot(int N, float *a, double *b, float no_conj)
+  {
+    double val = 0;
+    for (int i = 0; i < N; i++)
+      val += *a++ * *b++;
+    return val;
+  }
+
+  // versions for which we don't have Blas support
+  double dot(int N, double *a, float *b, float no_conj)
+  {
+    double val = 0;
+    for (int i = 0; i < N; i++)
+      val += *a++ * *b++;
+    return val;
+  }
+
+  // no_conj is normally 1, but -1 if the output needs to be conjugated.
+  std::complex<float> dot(int N, std::complex<float> *a, std::complex<float> *b,
+                          float no_conj)
+  {
+    float temp[2];
+    cblas_cdotc_sub(N, b, 1, a, 1, &temp);
+    return std::complex<float>(temp[0], no_conj * temp[1]);
+  }
+
+  std::complex<double> dot(int N, std::complex<double> *a,
+                           std::complex<double> *b, float no_conj)
+  {
+    double temp[2];
+    cblas_zdotc_sub(N, b, 1, a, 1, &temp);
+    return std::complex<double>(temp[0], no_conj * temp[1]);
+  }
+
+  template <class A, class B, typename U>
+  types::ndarray<typename A::dtype, types::pshape<long>>
+  do_correlate(A const &inA, B const &inB, U type, int out_inc)
+  // out_inc is used to indicate the inputs were swapped, which means that the
+  // output must be time reversed and conjgated
+  {
+    auto shapeA = sutils::array(inA.shape());
+    auto shapeB = sutils::array(inB.shape());
+
+    long NA = shapeA[0];
+    long NB = shapeB[0];
+
+    typedef typename decltype(inA + inB)::dtype out_type;
+    // At this point, handling views would slow things down tremendously
+    auto inA_ = functor::asarray{}(inA);
+    auto inB_ = functor::asarray{}(inB);
+    auto *ina_pt = (typename A::dtype *)inA_.buffer;
+    auto *inb_pt = (typename B::dtype *)inB_.buffer;
+
+    auto outN = 0;
+    if (type == "full")
+      outN = NA + NB - 1;
+    else if (type == "valid")
+      outN = NA - NB + 1;
+    else if (type == "same")
+      outN = NA;
+    types::ndarray<out_type, types::pshape<long>> out = {outN, out_type()};
+    out_type *out_ptr = (out_type *)out.buffer;
+
+    // For small correlations, numpy uses small_correlate, far more efficient.
+    // /Users/jlaroche/temp/NumpySrc/numpy/core/src/multiarray/arraytypes.c.src
```

Comment:
why don't you use the same strategy here? Do you prefer implementing that in a further PR?

Reply:
I looked at their implementation, it's pretty gnarly. It would probably be best to copy it straight from numpy but that also requires a bit of editing because of types etc...